### ISSUE TITLE
Amazon Cloud Player -- Selections open in browser

### ISSUE THREAD

**Issue by xmbwd** [tiliado/nuvola-app-amazon-cloud-player#13]
All of a sudden, when I click any of the selections in the player instead of opening the music selection in the player, it opens in a browser.  That is, any selection (e.g., Artists or Play list) does not have a result inside the player, but is instead routed to the default web browser.

**Comment by semidark** [comment]
Same for me. I attached the Console Debug Output of the Application Start and one "Click" wich opens in Firefox

[nuvolaplayer3_debug_amazon.txt](https://github.com/tiliado/nuvola-app-amazon-cloud-player/files/223642/nuvolaplayer3_debug_amazon.txt)

**Comment by WillGresham** [comment]
Also got the same issue, although interestingly right-click and then selecting 'Open Link' seems to open the links in the application rather than in the web browser.

**Comment by semidark** [comment]
@WillGresham Right click workaround works for me too

**Comment by jiri-janousek** [comment]
Hello everybody. Thanks for taking the time to report this issue. It seems Amazon changed addresses of various pages so the URL filter needs to be updated.

@SteVwonder, could you look at it? Thanks.

**Comment by SteVwonder** [comment]
Yes.  I can look at it this weekend.

**Comment by SteVwonder** [comment]
So I think I have a new working version. If someone else would be willing to test it out, that would be awesome.  https://github.com/SteVwonder/nuvola-app-amazon-cloud-player/tree/new_url

**Comment by chaosblog** [comment]
I checked the [new_url branch](/SteVwonder/nuvola-app-amazon-cloud-player/tree/new_url) and it works for me. :+1:

Before I had the same issue, the links were opened in a browser and were only usable in the application when right-clicked.
With the new_url code everything worked fine again. Thank you very much.

**Comment by SteVwonder** [comment]
Sweet. I setup a pull request to merge it into the main fork.  The only problem with the new version is that when you are listening to prime radio, there is no way to grab the album for scrobbling.  I was unable to find this information anywhere in the html, so I think its a limitation of their layout redesign.

@fenryxo, does this warrant a bump in the minor version?

**Comment by jiri-janousek** [comment]
> does this warrant a bump in the minor version?

Minor version changes only when there is a new release. Would you like me to create new release of your script? If yes, [create an issue for me to do it](https://tiliado.github.io/nuvolaplayer/development/apps/guidelines.html#releases) but don't change version numbers.

**Comment by chaosblog** [comment]
@SteVwonder when listening to a playlist or to prime radio, there is no album information.
Is it possible, to follow links and grep the new page? I haven't looked into the code of nuvola and the cloud player yet
When you can open the link to get the current song from the shop, you can get the album which contains the track.

**Comment by SteVwonder** [comment]
That is a good question. @fenryxo could probably answer in more detail.   My guess is that it is technically feasible (a hackish solution would be to just make the requests with javascript and parse it from there), but I also guess that it is principally in-feasible.  Without careful coding, this would result in an HTTP request to Amazon even time that Nuvola polls the plugin for song metadata (I am unsure of how often this occurs but I believe it is quite frequent).

And before we jump through hoops trying to grab and parse other html pages, does someone else mind digging into the Amazon site's HTML?  I was unable to find the album information anywhere (even with a ctrl-f search), but maybe someone else is cleverer than I am and can find the album information buried somewhere.

**Comment by jiri-janousek** [comment]
> Nuvola polls the plugin for song metadata (I am unsure of how often this occurs but I believe it is quite frequent).

Nuvola Core doesn't poll for metadata at all, it is just waiting for your script to send the metadata via `player.setTrack(track)` ([code](https://github.com/tiliado/nuvola-app-amazon-cloud-player/blob/master/integrate.js#L146)). Actually, it is your script which pools for metadata via the `WebApp.update` function scheduled to be executed every 500 ms ([code](https://github.com/tiliado/nuvola-app-amazon-cloud-player/blob/master/integrate.js#L165)).

>  Without careful coding, this would result in an HTTP request to Amazon even time that [script] polls the [page] for song metadata [...]

You should send the HTTP request only when track really changes. Store track metadata in WebApp object and compare them with freshly parsed data in next `WebApp.update()` run.

> And before we jump through hoops trying to grab and parse other html pages, does someone else mind digging into the Amazon site's HTML?

I might get to it later today.

**Comment by jiri-janousek** [comment]
> > And before we jump through hoops trying to grab and parse other html pages, does someone else mind digging into the Amazon site's HTML?
> 
> I might get to it later today.

Actually, I won't because I don't have Prime Radio subscription.

**Comment by chaosblog** [comment]
I digged into the HTML and found a place where the album name is hidden - when you listen to a playlist.

There are serveral tables for the playlist - 19 Tracks are in one table. The tables have an id, starting with 0 for the first table.

``` html
<table id="0" class="listPage" ...></table>
<table id="1" class="listPage" ...></table>
```

The current played song in the playlist table has an svg element and a span element which helps to find the current song - there is a td element afterwards which has the album name in it.

``` html
<tr class="playlistDetailsListItem" style="height:50px;">
    <td class="trackNumber ">
        <div class="listViewTrackNumber">
            <svg class="equalizer PLAYING"></svg>
            <span role="button" tabindex="0" aria-label="Abspielen" class="playingButton playerIconPlay">/span>
        </div>
    </td>
    <td class="trackAlbumArt ">
        <a class="" aria-label="HERE IS THE ALBUM NAME" href="album Detail URL">
            <span><div class="albumArtWrapper"><img src="Album Art URL" class="renderImage1" style="opacity: 1;"></div></span>
        </a>
    </td>
```

If playback is paused at the moment, then the svg and the span have other classes:

``` html
<svg class="equalizer PLAYING"></svg> <span role="button" tabindex="0" aria-label="Pause" class="playingButton playerIconPause"></span></div>
```

This only works for playlists. For radios I have to dig again and try to find the album name somewhere else

**Comment by SteVwonder** [comment]
It looks like Amazon updated their layout to include album names in the actual text (displayed to the user), so assuming you don't navigate away from your playlist page, we can parse the album name.  I have a prototype at [new_url_2](/SteVwonder/nuvola-app-amazon-cloud-player/tree/new_url_2).  @chaosblog, can you give the new branch a spin?  It should always produce the correct album name (assuming you don't navigate away).  If that works for you, I'll setup a PR into the master branch.

**Comment by jiri-janousek** [comment]
Hello @SteVwonder. This issue was opened more than 3 months ago. Are you going to push a fix to the master branch anytime soon?

**Comment by jiri-janousek** [comment]
I've adopted this abandoned script and released version 5.2 with fixes.

-- Jiri Janousek, Nuvola Player developer • _My work on Nuvola Player and user support are [funded by Nuvola Patrons](https://tiliado.eu/nuvolaplayer/funding/). [:heart: Make a donation](https://tiliado.eu/nuvolaplayer/funding/) to keep the project going. Thank you._
